lib/knapsack_pro/store/server.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+module KnapsackPro
+  module Store
+    class Server
+      extend Forwardable
+
+      def self.reset
+        stop
+        set_store_server_uri(nil)
+        @assigned_store_server_uri = nil
+        @client = nil
+      end
+
+      def self.start
+        return unless @server_pid.nil?
+
+        assign_available_store_server_uri
+
+        @server_pid = fork do
+          server_is_running = true
+
+          Signal.trap("TERM") {
+            server_is_running = false
+          }
+
+          DRb.start_service(store_server_uri, new)
+
+          while server_is_running
+            sleep 0.1
+          end
+
+          DRb.stop_service
+
+          # Wait for the drb server thread to finish before exiting.
+          DRb.thread&.join
+        end
+
+        ::Kernel.at_exit do
+          stop
+        end
+      end
+
+      def self.client
+        return @client unless @client.nil?
+
+        # must be called at least once per process
+        # https://ruby-doc.org/stdlib-2.7.0/libdoc/drb/rdoc/DRb.html
+        DRb.start_service
+
+        @client = DRbObject.new_with_uri(store_server_uri)
+
+        begin
+          retries ||= 0
+          @client.ping
+          @client
+        rescue DRb::DRbConnError
+          wait_seconds = 0.1
+          retries += wait_seconds
+          sleep wait_seconds
+          retry if retries <= 3 # seconds
+          raise
+        end
+      end
+
+      def_delegators :@queue_batch_manager, :add_batch, :last_batch_passed!, :last_batch_failed!, :batches
```

Comment:
We could, but I wanted to have a single DRb serivce (store) in a long term when we add more features to it (perhaps we would store data in memory instead of using json files). I prefer that instead of creating a new DRb service per feature.

I use the following architecture:

`KnapsackPro::Store::Client` - it exposes public methods that gem users can use. The API (public methods) should be stable so if we add new features they stay backward compatible.

`KnapsackPro::Store::Sever` - it's a DRb service that holds the data state in memory in a forked process (child process of the main process) so that we could share a state between processes like Knapsack process and Cucumber child process (Queue Mode).

`KnapsackPro::Store::QueueBatchManager` - a specific use case (a feature) . It holds a state of data for a given feature inside of the `KnapsackPro::Store::Sever` forked process.